### ISSUE TITLE
Understanding How Masking Affects BCE Loss Function

### ISSUE THREAD

**Issue by ymodak** [keras-team/keras#14987]
Cross posted from : https://github.com/tensorflow/tensorflow/issues/50710
Author : @chethanjjj

Hi, I'm experimenting with masking and have come across an interesting observation. I've created a simple GRU-based model with zeroed-out weights and biases along with a simple X and Y tensor. Theoretically, because my weights and biases are set to zero I should get the same loss value for the same dataset regardless if masking is on or not. However, I find that masking affects my results and I'd like more clarification on why this is happening. Below is my code to reproduce the difference. Additionally, I am using version 2.4.1 and 2.4.0 for tensorflow and keras, respectively

```
import tensorflow as tf
from tensorflow.keras import layers
from tensorflow.keras import losses
from tensorflow.keras import regularizers
from tensorflow.keras import initializers
from tensorflow.keras import Model
from tensorflow.keras import backend as K
from tensorflow.keras.optimizers import Adam

GRU_SIZE = 50
DENSE_DIM = 200

# generate data
data = tf.random.uniform([2, 4, 4], maxval=2, dtype=tf.int32, seed=1)
data = tf.cast(data, tf.float32)
data = K.eval(tf.identity(data))

# zero-out specific timesteps in each patient
data[0, 2, :] = 0.0
data[0, 3, :] = 0.0
data[1, 2, :] = 0.0
data[1, 3, :] = 0.0

# use the 1st n-1 timesteps in X to predict the last n-1 timesteps in Y
x_dummy_data = data[:, 0:-1, :]
y_dummy_data = data[:, 1:, :]

# create the masked model
input_shape = (x_dummy_data.shape[1], x_dummy_data.shape[2])
inputs = layers.Input(shape=input_shape)
masking = layers.Masking(mask_value=0, input_shape=input_shape)
input_masked = masking.compute_mask(inputs)
output1 = layers.Dense(DENSE_DIM)(inputs)
output3 = layers.GRU(GRU_SIZE, return_sequences=True)(output1, mask=input_masked)
output4_code = layers.Dense(input_shape[1], 
                                                   activation='sigmoid', 
                                                   name="output_code")(output3)
model_mask = Model(inputs=inputs, outputs = output4_code)
model_mask.summary()
model_mask.compile(loss=tf.keras.losses.BinaryCrossentropy(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE, from_logits=False),
                                     optimizer=Adam(learning_rate=1e-4))

# set weights and bias to zero
for ix, layer in enumerate(model_mask.layers):
     weights = layer.get_weights()
     for arr in weights:
        arr[(arr > 0) | (arr < 0)] = 0
     layer.set_weights(weights) 

# create the unmasked model
input_shape = (x_dummy_data.shape[1], x_dummy_data.shape[2])
inputs = layers.Input(shape=input_shape)
output1 = layers.Dense(DENSE_DIM)(inputs)
output3 = layers.GRU(GRU_SIZE, return_sequences=True)(output1)
output4_code = layers.Dense(input_shape[1], 
                                                   activation='sigmoid', 
                                                   name="output_code")(output3)
model_unmask = Model(inputs=inputs, outputs = output4_code)
model_unmask.summary()
model_unmask.compile(loss=tf.keras.losses.BinaryCrossentropy(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE, from_logits=False),
                                     optimizer=Adam(learning_rate=1e-4))

# set weights and bias to zero
for ix, layer in enumerate(model_unmask.layers):
     weights = layer.get_weights()
     for arr in weights:
        arr[(arr > 0) | (arr < 0)] = 0
     layer.set_weights(weights) 

# evaluate on fake data
model_unmask.evaluate(x_dummy_data, y_dummy_data)
model_mask.evaluate(x_dummy_data, y_dummy_data)
```

The model_unmask.evaluate leads to a loss of 0.6931471824645996, while the masked model is 0.4620981216430664. Can you explain how the latter is calculated? If weights and biases are set to 0 and the last layer is sigmoid activated then the only input into the loss function is a tensor of 0.5's. Shouldn't these both be the same when you take an average over the batch?

**Comment by tomerk** [comment]
From running the above in a colab:
```
> model_mask(x_dummy_data)
```
Outputs:
```
<tf.Tensor: shape=(2, 3, 4), dtype=float32, numpy=
array([[[0.5, 0.5, 0.5, 0.5],
        [0.5, 0.5, 0.5, 0.5],
        [0.5, 0.5, 0.5, 0.5]],

       [[0.5, 0.5, 0.5, 0.5],
        [0.5, 0.5, 0.5, 0.5],
        [0.5, 0.5, 0.5, 0.5]]], dtype=float32)>
```

And the mask is `model_mask(x_dummy_data)._keras_mask`:
```
<tf.Tensor: shape=(2, 3), dtype=bool, numpy=
array([[ True,  True, False],
       [ True,  True, False]])>
```

My guess is the mask gets incorporated in the loss reduction/computation, but it doesn't look like the guide on masks (https://keras.io/guides/understanding_masking_and_padding/) covers loss computation with masks.

@fchollet should be able to provide more details

**Comment by chethanjjj** [comment]
hi, any update on this issue?

**Comment by chethanjjj** [comment]
Hi, was this resolved or not?
